### ISSUE TITLE
Bug report: `event_loop_until_condition` raises when `ConditionTimeoutError` event loop terminated by something else

### ISSUE THREAD

**Issue by corranwebster** [enthought/pyface#1181]
## Environment

Toolkit: _Qt_

## Description

With #1168 no longer checking the `condition` after event loop termination, we have a number of tests in downstream projects which are now failing because the event loop terminates via a mechanism other than (a) the condition evaluating to `True` in the event loop, or (b) timing out; but after the event loop has finished the condition evaluates to `True`.

When this happens code generates a `ConditionTimeoutError`, which is misleading because there was no timeout.

While this may sometimes (or even most of the time) indicate an issue with the test or other code, there could be situations where the exit via another path can happen via an issue with underlying code that the tester doesn't control, or may even be intentional.  In any case, it feels like #1168 may also be too large a behaviour change for a bugfix release.

I think it may make sense to do a check for the case: `condition_result` is False, loop didn't time out, and `condition` evaluates to `True` and _warn_ instead of raising.   Whatever the fix is, it certainly shouldn't raise a `ConditionTimeoutError` in this case.

**Comment by mdickinson** [comment]
Agreed, with one suggested modification: one of the issues that was solved by the earlier PR was not to evaluate the condition _outside_ the scope of the event loop; I think we want to keep that fix. So I'd suggest dropping the `if condition()` clause from your warning, and warn unconditionally if we're exiting the event loop as a result of neither the timeout nor the condition becoming true.

**Comment by mdickinson** [comment]
And if we're changing this again, I'd also suggest initialising `condition_result` to `None` rather than `False`, just so that in interactive debugging it's possible to distinguish the case where the condition_result was never set (e.g., because calling the condition raised an exception) from the case where the condition was successfully evaluated and gave a result of `False`.

**Comment by corranwebster** [comment]
> one of the issues that was solved by the earlier PR was not to evaluate the condition _outside_ the scope of the event loop; I think we want to keep that fix

I was hoping by guarding with the initial check of `condition_result` we'd avoid the problematic cases that #1168 was trying to fix: `condition()` is only called in special cases and avoids the "evaluated True in loop by failed at the end".

But warning unconditionally is probably a reasonable thing to do if only for simplicity.  Tests which care can potentially double-check the condition if appropriate.

I agree about using `None` - the warning message should then probably differ based on whether the condition was ever evaluated.
